internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,606 @@
+package pam_test
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/authd"
+	"github.com/ubuntu/authd/cmd/authd/daemon"
+	"github.com/ubuntu/authd/internal/testutils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+var configFmt = `
+systemdirs:
+    socketpath: %s
+    brokercfgdir: %s
+`
+
+var (
+	fixturesPath = filepath.Join("testdata", "fixtures")
+)
+
+func TestAvailableBrokers(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		cfgDir string
+
+		wantErr bool
+	}{
+		"All brokers when config dir only has valid brokers":      {cfgDir: "valid_brokers"},
+		"Only correct brokers when config dir has mixed brokers":  {cfgDir: "mixed_brokers"},
+		"Only local broker when config dir only has invalid ones": {cfgDir: "invalid_brokers"},
+		"Only local broker when config dir does not exist":        {cfgDir: "does/not/exist"},
+		"No error even if broker is not exported on dbus":         {cfgDir: "not_on_bus"},
+
+		"Error when broker config dir is a file": {cfgDir: "file_config_dir", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			client := startClient(t, filepath.Join(fixturesPath, tc.cfgDir))
+
+			abResp, err := client.AvailableBrokers(context.Background(), &authd.Empty{})
+			if tc.wantErr {
+				require.Error(t, err, "AvailableBrokers should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "AvailableBrokers should not return an error, but did")
+			got := abResp.GetBrokersInfos()
+			want := testutils.LoadWithUpdateFromGoldenYAML(t, got)
+			require.Equal(t, want, got, "AvailableBrokers returned unexpected brokers")
+		})
+	}
+}
+
+func TestSetDefaultBrokerForUser(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username string
+		noBroker bool
+
+		wantErr bool
+	}{
+		"Set default broker for existing user": {username: "user"},
+
+		"Error when username is empty":     {wantErr: true},
+		"Error when broker does not exist": {username: "user", noBroker: true, wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			client := startClient(t, filepath.Join(fixturesPath, "mixed_brokers"))
+
+			abResp, err := client.AvailableBrokers(context.Background(), &authd.Empty{})
+			require.NoError(t, err, "Setup: failed to get available brokers for tests")
+
+			//nolint:gosec // This just to vary the selected broker in the tests, we don't care about randomness safety.
+			wantID := abResp.GetBrokersInfos()[rand.Intn(len(abResp.GetBrokersInfos()))].GetId()
```

Comment:
Let’s not introduce randomness in the vbroker selection, but ensure that we always select the same for a given tests. This way, if we have an error, we are sure that we can reproduce and not pass an additional amount of time wondering why the selected broker is always different…